docs/modules/ROOT/pages/servlet/authorization/method-security.adoc

```diff
@@ -900,6 +900,127 @@ open class BankService {
 
 This results in more readable method definitions.
 
+You can also add parameters to your custom annotations.
+For example, instead of `@IsAdmin`, you can create something more generic like `@HasRole` like so:
+
+[tabs]
+======
+Java::
++
+[source,java,role="primary"]
+----
+@Target({ ElementType.METHOD, ElementType.TYPE })
+@Retention(RetentionPolicy.RUNTIME)
+@PreAuthorize("hasRole('${value}')")
+public @interface HasRole {
+	String value();
+}
```

Comment:
I wonder if using property placeholders would confuse users who may assume the placeholders can be sourced from JVM system properties, environment variables, Spring Boot application properties, etc.

Have you considered treating the text as a template containing SpEL expressions which can access the annotation's attributes?

Reply:
> Have you considered treating the text as a template containing SpEL expressions which can access the annotation's attributes?

One thing that I considered was using SpEL variables; however, I ran into trouble when an annotation-attribute-as-SpEL-variable referred to another SpEL variable:

```java
@PreAuthorize(hasRole(#value))
public @interface HasRole {
    String value();
}

// ...

@HasRole("#parameter")
public String doThis(String parameter) {
    // ...
}
```

I'm not sure of another way that the SpEL expression could access the annotation's attributes that would also allow for the attributes to be SpEL expressions themselves. Can you elaborate?

Reply:
Alternatively, I could change the delimiter, like so:

```java
@PreAuthorize("hasRole('{value}')")
public @interface HasRole {
	String value();
}
```

so that it matches the syntax for `@RequestMapping` URI templates.